src/mlpack/methods/reinforcement_learning/replay/random_replay.hpp

```diff
@@ -0,0 +1,142 @@
+/**
+ * @file random_replay.hpp
+ * @author Shangtong Zhang
+ *
+ * This file is an implementation of random experience replay.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RL_REPLAY_RANDOM_REPLAY_HPP
+#define MLPACK_METHODS_RL_REPLAY_RANDOM_REPLAY_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace rl {
+
+/**
+ * Implementation of random experience replay.
+ *
+ * @tparam EnvironmentType Desired task.
+ */
+template <typename EnvironmentType>
+class RandomReplay
+{
+ public:
+  using ActionType = typename EnvironmentType::Action;
+  using StateType = typename EnvironmentType::State;
+
+  /**
+   * Construct an instance of random experience replay class.
+   *
+   * @param batchSize # of examples returned at each sample.
+   * @param capacity Total memory size in terms of # of examples.
+   * @param dimension The dimension of an encoded state.
+   */
+  RandomReplay(size_t batchSize,
+               size_t capacity,
+               size_t dimension = StateType::dimension) :
+      batchSize(batchSize),
+      capacity(capacity),
+      position(0),
+      states(dimension, 0),
+      nextStates(dimension, 0)
+  { /* Nothing to do here. */ }
+
+  /**
+   * Store the given experience.
+   *
+   * @param state Given state.
+   * @param action Given action.
+   * @param reward Given reward.
+   * @param nextState Given next state.
+   * @param isEnd Whether next state is terminal state.
+   */
+  void Store(const StateType& state, ActionType action,
+             double reward, const StateType& nextState, bool isEnd)
+  {
+    if (isTerminal.n_elem < capacity)
```

Comment:
I'm not sure. Two days ago it's preallocated, but sometimes user may declare a pretty large capacity and actually only small part of the memory is used.  So I modify the code to dynamic allocation. Maybe I can add a parameter in the constructor to determine whether to preallocate or not.